subsys/bluetooth/controller/ll_sw/ull_sched.c

```diff
@@ -4,43 +4,536 @@
+#include "util/mayfly.h"
+
+#include "ticker/ticker.h"
 
 #include "pdu.h"
 
 #include "lll.h"
+#include "lll_vendor.h"
 #include "lll_scan.h"
+#include "lll_conn.h"
 
 #include "ull_scan_types.h"
+#include "ull_conn_types.h"
+
+#include "ull_internal.h"
+#include "ull_conn_internal.h"
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_HCI_DRIVER)
+#define LOG_MODULE_NAME bt_ctlr_ull_sched
+#include "common/log.h"
+#include "hal/debug.h"
+
+#if defined(CONFIG_BT_CTLR_CONN_PARAM_REQ)
+static void win_offset_calc(struct ll_conn *conn_curr, u8_t is_select,
+			    u32_t *ticks_to_offset_next, u16_t conn_interval,
+			    u8_t *offset_max, u8_t *win_offset);
+#endif /* CONFIG_BT_CTLR_CONN_PARAM_REQ */
+static void after_mstr_offset_get(u16_t conn_interval, u32_t ticks_slot,
+				  u32_t ticks_anchor, u32_t *win_offset_us);
+static void ticker_op_cb(u32_t status, void *param);
 
 void ull_sched_after_mstr_slot_get(u8_t user_id, u32_t ticks_slot_abs,
 				   u32_t *ticks_anchor, u32_t *us_offset)
 {
-	/* TODO: */
+	u32_t ticks_to_expire_prev;
+	u32_t ticks_slot_abs_prev;
+	u32_t ticks_to_expire;
+	u8_t ticker_id_prev;
+	u8_t ticker_id;
+
+	ticks_slot_abs += HAL_TICKER_US_TO_TICKS(EVENT_JITTER_US << 3);
+
+	ticker_id = ticker_id_prev = 0xff;
+	ticks_to_expire = ticks_to_expire_prev = *us_offset = 0U;
+	ticks_slot_abs_prev = 0U;
+	while (1) {
+		u32_t volatile ret_cb = TICKER_STATUS_BUSY;
+		struct ll_conn *conn;
+		u32_t ret;
+
+		ret = ticker_next_slot_get(TICKER_INSTANCE_ID_CTLR, user_id,
+					   &ticker_id, ticks_anchor,
+					   &ticks_to_expire, ticker_op_cb,
+					   (void *)&ret_cb);
+
+		if (ret == TICKER_STATUS_BUSY) {
+			while (ret_cb == TICKER_STATUS_BUSY) {
+				ticker_job_sched(TICKER_INSTANCE_ID_CTLR,
+						 user_id);
+			}
+		}
+
+		LL_ASSERT(ret_cb == TICKER_STATUS_SUCCESS);
+
+		if (ticker_id == 0xff) {
+			break;
+		}
+
+		if ((ticker_id < TICKER_ID_CONN_BASE) ||
+		    (ticker_id > TICKER_ID_CONN_LAST)) {
+			continue;
+		}
+
+		conn = ll_conn_get(ticker_id - TICKER_ID_CONN_BASE);
+		if (conn && !conn->lll.role) {
+			u32_t ticks_to_expire_normal = ticks_to_expire;
+			u32_t ticks_slot_abs_curr = 0;
+#if defined(CONFIG_BT_CTLR_LOW_LAT)
+#if defined(CONFIG_BT_CTLR_XTAL_ADVANCED)
+			if (conn->evt.ticks_xtal_to_start & XON_BITMASK) {
+				u32_t ticks_prepare_to_start =
+					MAX(conn->evt.ticks_active_to_start,
+					    conn->evt.ticks_preempt_to_start);
+
+				ticks_slot_abs_curr =
+					conn->evt.ticks_xtal_to_start &
+					~XON_BITMASK;
+				ticks_to_expire_normal -=
+					ticks_slot_abs_curr -
+					ticks_prepare_to_start;
+			} else
+#endif /* CONFIG_BT_CTLR_XTAL_ADVANCED */
+			{
+				u32_t ticks_prepare_to_start =
+					MAX(conn->evt.ticks_active_to_start,
+					    conn->evt.ticks_xtal_to_start);
+
+				ticks_slot_abs_curr = ticks_prepare_to_start;
+			}
+#endif
+
+			ticks_slot_abs_curr +=
+				conn->evt.ticks_slot +
+				HAL_TICKER_US_TO_TICKS(EVENT_JITTER_US << 3);
+
+			if ((ticker_id_prev != 0xFF) &&
```

Comment:
nitpick: elsewhere lower-case f is used
if 0xff is a flag, shouldn't we use a #define for this?